### PR TITLE
Ab Test partner

### PR DESCRIPTION
**Pull request by pm-shashank-jain**
Ab Test feature for Partners and timeout

### CODE REVIEW

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src_new/config.js  4
         

Clones added
============
- test/config.spec.js  3
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

File: src_new/conf.js [hunk 1]
```diff
@@ -11,19 +11,25 @@ exports.pwt = {
...
 	ccpa: "0",
 	ccpaCmpApi: "iab",
-	ccpaTimeout: "10000"
+	ccpaTimeout: "10000",
+	abTestEnabled:"0"
 };
 
 // singleImpression is used to enable feature of sending single impression for multiple size ad slot earlier there were multiple impression for multiple sizes
-
+exports.testConfigDetails = {
```
Comment (pm-shashank-jain) [inline src_new/conf.js:27]: Add a comment, as this is for test purpose only

File: src_new/config.js [hunk 2]
```diff
@@ -216,27 +217,56 @@ exports.getCCPACmpApi = function () {
...
-exports.isUsePrebidKeysEnabled = function(){
+exports.isUsePrebidKeysEnabled = function () {
 	// note: not using window.parseInt as this function is also used in build.sh that runs in NodeJS environment
 	return parseInt(config[CONSTANTS.CONFIG.COMMON][CONSTANTS.CONFIG.USE_PREBID_KEYS]) === 1;
 };
 
 exports.PBJS_NAMESPACE = config[CONSTANTS.CONFIG.COMMON][CONSTANTS.COMMON.PBJS_NAMESPACE] || "pbjs";
+
+exports.updateABTestConfig = function () {
+	if (refThis.isAbTestEnabled) {
```
Comment (pm-shashank-jain) [inline src_new/config.js:246]: Update to this as function

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src_new/config.js  5
         

Clones added
============
- test/config.spec.js  2
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

File: src_new/config.js [hunk 3]
```diff
@@ -216,27 +217,56 @@ exports.getCCPACmpApi = function () {
...
+		var randomNumberBelow100 = util.getRandomNumberBelow100();
+		var testGroupDetails = refThis.getTestGroupDetails();
+		if (testGroupDetails && testGroupDetails.testGroupSize && randomNumberBelow100 < testGroupDetails.testGroupSize) {
```
Comment (pm-shashank-jain) [inline src_new/config.js:249]: Add Comment if AB Test is enabled and what config is applied

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src_new/config.js  5
         

Clones added
============
- test/config.spec.js  2
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src_new/config.js  5
         

Clones added
============
- test/config.spec.js  2
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- src_new/config.js  5
         

Clones added
============
- test/config.spec.js  2
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- src_new/config.js  5
         

Clones added
============
- test/config.spec.js  2
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- src_new/config.js  7
         

Clones added
============
- test/config.spec.js  2
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

File: src_new/config.js (hunk 3 above)
Comment (pm-shashank-jain) [reply src_new/config.js:265]: Done

File: src_new/conf.js (hunk 1 above)
Comment (pm-shashank-jain) [reply src_new/conf.js]: Done

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src_new/config.js  7
         

Clones added
============
- test/config.spec.js  58
- src_new/conf.js  9
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

File: src_new/config.js [hunk 4]
```diff
@@ -261,18 +279,41 @@ exports.updatePWTConfig = function () {
...
+					testConfig[key] = controlConfig[key];
+				} else if (Object.keys(testConfig[key]).length > 0 && controlConfig[key] && Object.keys(controlConfig[key]).length > 0) {
+					testConfig[key] = refThis.getMergedConfig(testConfig[key], controlConfig[key]);
+				}
+			}
+		}
+		window.PWT.testGroupId = 1;
+		return testConfig;
+	} else{
+		return {};
```
Comment (pm-shashank-jain) [inline src_new/config.js:303]: Since there will be a single AB Test type only, we need to return control partners if there  are no test type

**Comment by pm-shashank-jain** [comment]
Add a flag for Hybrid Partners to be sent to pubmaticServerBidAdapter.js

File: test/config.spec.js [hunk 5]
```diff
@@ -1233,5 +1233,741 @@ describe('Config', function() {
...
+            };
+            var updatedPartners = CONFIG.updatePartnerConfig(CONF[CONSTANTS.COMMON.TEST_PARTNER],adapters);
+            expect(updatedPartners).to.be.deep.equal(expectedResult);
+            done();
+        });
+
+        it('should return empty object if test config is not present but test is enabled',function(done){
+            CONF[CONSTANTS.COMMON.TEST_PARTNER]= {};
+            var updatedPartners = CONFIG.updatePartnerConfig(CONF[CONSTANTS.COMMON.TEST_PARTNER],adapters);
+            expect(updatedPartners).to.be.deep.equal({});
```
Comment (pm-shashank-jain) [inline test/config.spec.js:1688]: update this to control partner

File: test/config.spec.js [hunk 6]
```diff
@@ -1233,5 +1233,741 @@ describe('Config', function() {
...
+                var updatedPartners = CONFIG.updatePartnerConfig(testIdentityPartners,identityPartners);
+                console.log("Updated Partners are ", JSON.stringify(updatedPartners));  
+                expect(updatedPartners).to.be.deep.equal(expectedResult);
+                done();
+            });
+
+            it('should return empty object if test config is not present but test is enabled',function(done){
+                testIdentityPartners = {};
+                var updatedPartners = CONFIG.updatePartnerConfig(testIdentityPartners, identityPartners);
+                expect(updatedPartners).to.be.deep.equal({});
```
Comment (pm-shashank-jain) [inline test/config.spec.js:1865]: expected result is control identity partners

File: src_new/config.js (hunk 4 above)
Comment (pm-shashank-jain) [reply src_new/config.js:303]: Done

File: test/config.spec.js (hunk 5 above)
Comment (pm-shashank-jain) [reply test/config.spec.js:1688]: Done

File: test/config.spec.js (hunk 6 above)
Comment (pm-shashank-jain) [reply test/config.spec.js:1865]: Done

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src_new/config.js  7
         

Clones added
============
- test/config.spec.js  58
- src_new/conf.js  9
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

**Comment by sureshghare** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src_new/config.js  7
         

Clones added
============
- test/config.spec.js  58
- src_new/conf.js  9
         
```


See the complete overview on [Codacy](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=14002&bid=47122)

**Comment by pm-shashank-jain** [comment]
PR Raised to merge in nightly #228

**Pull request closed without merging**